tools/mull-cxx-frontend/src/MullClangPlugin.cpp

```diff
@@ -0,0 +1,132 @@
+#include "ASTInstrumentation.h"
+#include "ASTMutationsSearchVisitor.h"
+#include "ASTNodeFactory.h"
+#include "MullASTMutator.h"
+#include "MutationMap.h"
+
+#include <clang/AST/AST.h>
+#include <clang/AST/ASTConsumer.h>
+#include <clang/Frontend/CompilerInstance.h>
+#include <clang/Frontend/FrontendPluginRegistry.h>
+#include <clang/Sema/Sema.h>
+#include <clang/Sema/SemaConsumer.h>
+#include <llvm/Support/raw_ostream.h>
+
+using namespace clang;
+using namespace llvm;
+
+namespace {
+
+class MullASTConsumer : public ASTConsumer {
+  CompilerInstance &Instance;
+  std::unique_ptr<MullASTMutator> astMutator;
+  MutationMap mutationMap;
+
+public:
+  MullASTConsumer(CompilerInstance &Instance, const MutationMap mutationMap)
+      : Instance(Instance), astMutator(nullptr), mutationMap(mutationMap) {}
+
+  void Initialize(ASTContext &Context) override {
+    ASTConsumer::Initialize(Context);
+  }
+
+  /// This function can be considered a main() function of the
+  /// mull-cxx-frontend plugin. This method is called multiple times by
+  /// clang::ParseAST() for each declaration when it's finished being parsed.
+  /// For each found function declaration below, a two-pass approach is used:
+  /// 1) First all mutation points are found in the function declaration by the
+  /// recursive AST visitor class ASTMutationsSearchVisitor.
+  /// 2) For each mutation point, the mutations are performed on the Clang AST
+  /// level. The mutation is performed by the higher-level MullASTMutator class
+  /// which class to the lower-level ClangASTMutator class.
+  bool HandleTopLevelDecl(DeclGroupRef DG) override {
+    /// Could be a better place to create this. But at Initialize(), getSema()
+    /// hits an internal assert because it is not initialized yet at that time.
+    if (!astMutator) {
+      astMutator = std::make_unique<MullASTMutator>(Instance.getASTContext(), Instance.getSema());
+      astMutator->instrumentTranslationUnit();
+    }
+
+    for (DeclGroupRef::iterator I = DG.begin(), E = DG.end(); I != E; ++I) {
+      if ((*I)->getKind() != Decl::Function) {
+        continue;
+      }
+
+      FunctionDecl *f = static_cast<FunctionDecl *>(*I);
+      if (f->getDeclName().getAsString() == "main") {
+        continue;
+      }
+
+      clang::SourceLocation functionLocation = f->getLocation();
+      if (Instance.getSourceManager().isInSystemHeader(functionLocation)) {
+        continue;
+      }
+      std::string sourceFilePath = Instance.getSourceManager().getFilename(functionLocation).str();
+      if (sourceFilePath.find("include/gtest") != std::string::npos) {
```

Comment:
Do we need this immediately? i.e. can we mutate everything and then add filters at a later stage?

Reply:
The same answer as for the previous comment - let's keep it until we have filters!

One of these filters filters gtest's translation units, the other one the inlined code, so they are both useful for now.